### PR TITLE
objstore/s3: Add list_objects_version to S3 config

### PR DESCRIPTION
**Pull request by sevagh**
Signed-off-by: Sevag Hanssian <sevagh@protonmail.com>

* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

I added an optional config "list_objects_version: string" to the S3 configuration.

## Verification

I have a bucket in production which requires ListObjectsV1 to work (I need to deploy a custom binary with minio UseV1 set to true). After making this PR, I tested "thanos bucket tools web" with the new config option and it works.

### CODE REVIEW

**Comment by sevagh** [comment]
Solves https://github.com/thanos-io/thanos/issues/3291

**Review by kakkoyun — APPROVED** [review]
LGTM. Thanks for your contributions.

Let's wait for another approval and then we can merge it.

**Comment by sevagh** [comment]
> ci/circleci: test — Your tests failed on CircleCI 
I can't see what failed here. Is it related to my code?

File: docs/storage.md
Comment (bwplotka) [inline docs/storage.md:114]: I would vote for `list_objects_version: "v1"` with default being `"v2"`. It might be much more flexible, WDYT? (:

**Review by bwplotka — COMMENTED** [review]
Thanks! One suggestion for flexibility and LGTM (:

File: docs/storage.md
Comment (sevagh) [reply docs/storage.md:114]: So @bwplotka should I make the switch? I'll defer to your opinion.

My only preference for this PR is that it gets merged into main sooner rather than later so I can deploy a real upstream Thanos (and not my own binary :))

Comment (bwplotka) [reply docs/storage.md:114]: Yes please! Let's switch. It's a good practice overall for the future (what if we change S3 client to different one?) 

That's the last suggestion, I think once it's done we are good to merge :hugs:

Comment (sevagh) [reply docs/storage.md:114]: Question. I am choosing to set it as a string in the `type Config struct`, and leave it as a bool in `type Bucket string` (which contains a minio.Client as a field, so it makes sense to have a minio-specific in that struct).

The function NewBucketWithConfig is a good place to store the logic of parsing `""|"v2"` or `v1` into the appropriate bool value. Should I break the parsing logic into a function so I can unit test it?
```
var useListObjectsV1 bool
if config.ListObjectsVersion == "" || config.ListObjectsVersion == "v2" {
    useListObjectsV1 = false
} else if config.ListObjectsVersion == "v1" {
    useListObjectsV1 = true
} else {
    listObjectsErrMsg := errors.Errorf("Unsupported version %q was provided. Supported values are v1, v2", config.ListObjectsVersion)
    return nil, errors.Wrap(listObjectsErrMsg, "Initialize s3 client list objects version")
}
```

Comment (sevagh) [reply docs/storage.md]: It's more clear if I push commit (just did).

File: pkg/objstore/s3/s3.go
Comment (sevagh) [inline pkg/objstore/s3/s3.go:260]: I'm wondering if this code needs a unit test.

Comment (bwplotka) [reply pkg/objstore/s3/s3.go:260]: Probably, but also this code can be simplified to:

```
	if config.ListObjectsVersion =! "" && c config.ListObjectsVersion =! "v1" && config.ListObjectsVersion != "v2" {
		return nil, errors.Errorf(Initialize s3 client list objects version: Unsupported version %q was provided. Supported values are v1, v2", config.ListObjectsVersion)
       }
       return Bucket{ 
           .....
           listObjectsV1:  config.ListObjectsVersion == "v1",
}
```

**Review by bwplotka — APPROVED** [review]
Small suggestion and LGTM (:

File: pkg/objstore/s3/s3.go
Comment (sevagh) [reply pkg/objstore/s3/s3.go:254]: Nice one, wish I wrote it that way the first time.

**Comment by sevagh** [comment]
I rebased on the tip of master to fix the CHANGELOG conflict. This should be OK to merge if it passes the checks?

File: CHANGELOG.md
Comment (bwplotka) [inline CHANGELOG.md:28]: Hm, I think rebase went wrong here?

**Review by bwplotka — APPROVED** [review]
LGTM, just one thing - rebase went wrong?

File: CHANGELOG.md
Comment (sevagh) [reply CHANGELOG.md:28]: The CHANGELOG looks different from the time I had to modify it: https://github.com/thanos-io/thanos/commit/57076a58680c8a318aefa2630ec6319f549acd56#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4ed

This has me confused, TBH.

Comment (sevagh) [reply CHANGELOG.md]: I just fixed it manually.

**Comment by bwplotka** [comment]
Thanks!

**Pull request merged**
